test/libsolidity/util/compiler/InternalCompiler.cpp

```diff
@@ -0,0 +1,180 @@
+    m_stack.setEOFVersion(_input.eofVersion);
+}
+
+std::pair<MappedContracts, Errors> InternalCompiler::internalCompile(
+    CompilerInput const& _input
+)
+{
+    m_stack.compile();
+
+    MappedContracts mappedContracts{};
+    for (auto sourceName: m_stack.sourceNames())
+    {
+        std::vector<CompiledContract> compiled;
+        for (auto contract: m_stack.contractDefinitions(sourceName))
+        {
+            std::string contractName = contract->fullyQualifiedName();
+            LinkerObject object = m_stack.object(contractName);
+            LinkerObject runtimeObject = m_stack.runtimeObject(contractName);
+            bool hasUnlinkedReferences = !object.linkReferences.empty();
+            bytes cborMetadata = m_stack.cborMetadata(contractName);
+
+            std::optional<AssemblyItems> assemblyItems;
+            std::optional<AssemblyItems> runtimeAssemblyItems;
+
+            if (!_input.eofVersion.has_value())
+            {
+                if (auto items = m_stack.assemblyItems(contractName); items != nullptr)
+                    assemblyItems = std::make_optional(*items);
+
+                if (auto items = m_stack.runtimeAssemblyItems(contractName); items != nullptr)
+                    runtimeAssemblyItems = std::make_optional(*items);
+            }
+
+            if (!m_stack.isExperimentalSolidity())
+                compiled.emplace_back(CompiledContract{
+                    contractName,
+                    object.bytecode,
+                    runtimeObject.bytecode,
+                    hasUnlinkedReferences,
+                    cborMetadata,
+                    assemblyItems,
+                    runtimeAssemblyItems,
+                    m_stack.metadata(contractName),
+                    m_stack.contractABI(contractName),
+                    m_stack.interfaceSymbols(contractName),
+                    generateEventSignatures(contractName)
+                });
+            else
+                compiled.emplace_back(CompiledContract{
+                    contractName,
+                    object.bytecode,
+                    runtimeObject.bytecode,
+                    hasUnlinkedReferences,
+                    cborMetadata,
+                    assemblyItems,
+                    runtimeAssemblyItems,
+                    std::nullopt,
+                    std::nullopt,
+                    std::nullopt,
+                    std::vector<AnnotatedEventSignature>{}
+                });
+        }
+        mappedContracts.insert(std::make_pair(sourceName, compiled));
+    }
+
+    return std::make_pair(mappedContracts, m_stack.errors());
+}
+
+std::vector<AnnotatedEventSignature> InternalCompiler::generateEventSignatures(
+    std::string const& _contractName
+) const
+{
+    std::vector<AnnotatedEventSignature> signatures;
+    ContractDefinition const& contract = m_stack.contractDefinition(_contractName);
+
+    for (EventDefinition const* event: contract.events() + contract.usedInterfaceEvents())
```

Comment:
you can use `ranges::views::concat` here to avoid creating an entirely new vector